### PR TITLE
Fix FSDP checkpoint conversion and loading for Qwen3.5-VL

### PR DESCRIPTION
**Pull request by DAISY-gh**
# What does this PR do ?
Experimenting with Qwen3.5-VL training, comparing EP-only versus FSDP+EP setups. The FSDP+EP setup encountered several checkpoint conversion and loading issues. Here is a concise summary of the fixes included in this PR:

1. Granular SWiGLU Splitting Problem: Global SWiGLU logic falsely split Vision Encoder's GELU fc1 weights (causing loss spikes: ~1.5 vs expected ~0.7). The regex also missed biases.
Fix: Added per-TransformerLayer gated_linear_unit check in fsdp_dtensor_checkpoint.py. Added --swiglu-modules flag and updated bias regex (r'((?:weight|bias)\d*)(.*)') in checkpoint_inspector.py. Purpose: Restrict SWiGLU splitting strictly to LM layers, protecting Vision Encoder initialization.

2. GDN (Gated DeltaNet) Key Merging Problem: torch_dist stores GDN sub-keys separately, but FSDP expects a single merged tensor.
Fix: Added --merge-gdn flag to checkpoint_inspector.py. Purpose: Align GDN composite tensor structures with FSDP expectations.

3. MTP Layer Key Renaming Problem: torch_dist MTP key paths use transformer_layer, but FSDP expects mtp_model_layer.
Fix: Added --rename-mtp-keys flag to checkpoint_inspector.py. Purpose: Resolve MTP path naming mismatches during checkpoint conversion.

image
image

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: tools/checkpoint/checkpoint_inspector.py [hunk 1]
```diff
@@ -597,22 +699,95 @@ def has_layer_index(key: str) -> bool:
...
 
+    if _swiglu_split_count > 0:
+        rank0_echo(f"[SWiGLU] Split {_swiglu_split_count} fc1 keys into _w/_v pairs.")
+    elif _swiglu_prefixes is not None and len(_swiglu_prefixes) > 0:
+        rank0_echo("[SWiGLU] WARNING: modules specified but 0 keys were split — check module names.")
+
+    # Merge GDN (Gated DeltaNet) composite keys.
+    # The torch_dist format stores in_proj.weight and conv1d.weight as separate
+    # sub-tensors (.query, .key, .value, .z, .beta, .alpha), but the FSDP model
+    # expects them as single concatenated tensors.
```
Comment (shjwudp) [inline tools/checkpoint/checkpoint_inspector.py]: Hi Daisy, thanks for submitting the fix.
The GDN weight might be formed by concatenating tensors from N TP shards. In that case, when FSDP-TP training is enabled, loading a checkpoint with a different TP parallel mapping could cause errors. We might need to handle the GDN weights similar to how we did with SwiGLU — by splitting the tensor buffer back into the original TP tensors and wrapping them as DTensors. Would you be interested in helping implement this?

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py [hunk 2]
```diff
@@ -321,14 +382,191 @@ def split_swiglu_linear_fc1(data, dist_param, swiglu_shard_axis, is_expert_param
...
                     )
-                    # Update the optimizer state dict with the new keys
                     new_opt_state_dict[f"{key}_w"][subkey] = weight_w
                     new_opt_state_dict[f"{key}_v"][subkey] = weight_v
             optimizer_state_dict["state"] = new_opt_state_dict
 
     return model_state_dict, optimizer_state_dict
 
 
+def handle_gdn_in_state_dict(model, model_state_dict, optimizer_state_dict):
```
Comment (shjwudp) [inline megatron/core/transformer/fsdp_dtensor_checkpoint.py:390]: Great, thanks. Can we cancel the Merge-GDN flag now? AKA checkpoint_inspector.py#711. 🤔

Comment (DAISY-gh) [reply megatron/core/transformer/fsdp_dtensor_checkpoint.py:390]: Removed the flag, thanks for pointing out.

**Comment by DAISY-gh** [comment]
> Looks good to me! Really appreciate your help expanding and fixing the checkpoint compatibility with the new MCore module (GDN). I've approved the PR, but you might need to add some tests to get the CI to pass.

Added the test.

**Review by yaoyu-33 — APPROVED** [review]

**Comment by cspades** [comment]
/ok to test [16c920d](https://github.com/NVIDIA/Megatron-LM/pull/3936/commits/16c920d525b49b24aa3c5322fffcdf5f339427ae)

**Comment by cspades** [comment]
Hey @DAISY-gh looks like your PR has a linting error, could you run the linter in `./tools/autoformat.sh`? Thanks!

**Comment by DAISY-gh** [comment]
> Hey @DAISY-gh looks like your PR has a linting error, could you run the linter in `./tools/autoformat.sh`? Thanks!

Fixed and it passed now

**Comment by cspades** [comment]
/ok to test [c4b82d9](https://github.com/NVIDIA/Megatron-LM/pull/3936/commits/c4b82d9eafee3ac04c071f43f9f1877e46a75fe0)

**Comment by cspades** [comment]
/ok to test [2b58f15](https://github.com/NVIDIA/Megatron-LM/pull/3936/commits/2b58f15866f4bdb585ac51d19e27877d1aa6f7df)

**Comment by cspades** [comment]
Something wonky with the CI/CD test, failed even though everything passed. Rebased + rerunning.
